libgalois/include/katana/GraphTopology.h

```diff
@@ -1428,6 +1430,79 @@ class KATANA_EXPORT PGViewCache {
 KATANA_EXPORT GraphTopology CreateUniformRandomTopology(
     const size_t num_nodes, const size_t edges_per_node) noexcept;
 
+template <bool IS_SYMMETRIC = false>
+class KATANA_EXPORT TopologyBuilderImpl : public GraphTopologyTypes {
+  using AdjVec = std::vector<Node>;
+
+  std::vector<AdjVec> all_nodes_adj_;
+
+  bool IsValidNode(Node id) const noexcept {
+    return id < all_nodes_adj_.size();
+  }
+
+  void AddEdgeImpl(Node src, Node dst) noexcept {
+    KATANA_LOG_DEBUG_ASSERT(IsValidNode(src));
+    auto& adj_list = all_nodes_adj_[src];
+
+    [[maybe_unused]] bool not_found =
+        (std::find(adj_list.begin(), adj_list.end(), dst) == adj_list.end());
+    KATANA_LOG_DEBUG_ASSERT(not_found);
```

Comment:
any particular reason multi-edges aren't allowed, or just a decision you made?

Reply:
We can enable multi-edges too (via a separate flag). I just didn't do it for now. 
CC: @mindv0rtex